### PR TITLE
Unified test runner engine

### PR DESCRIPTION
**Pull request by dorchard**
This PR provides an interface that will allow us to target our test suite at both Python and MATLAB code.

The essential idea is captured by the abstract base class `TestEngine` which describes an interface with methods:
* `_repr_pretty` which just needs to return a string
* `convert` which takes some input in Python/numpy standard data types and converts it to the format required by the runner
* `equal` which tests if two values are queal in the representation used by the runner.

There are then two classes provided inheriting from this class `PythonEngine` and `MatlabEngine`. The `PythonEngine` largely does not very much, but given a runner `r` then `r.funname(x,y)` will call a Python function `funname` in the global scope with inputs `x` and `y`. The `MatlabEngine` passes any such function calls to the MATLAB Engine backend.

-----------------
Update: 

In order to call the Python code from the tests we have to explicitly import the relevant modules in `conftest.py` as demonstrated here so far (see e.g., https://github.com/Cambridge-ICCS/ONEFlux/blob/unifiedEngine/tests/conftest.py#L34-L35) 

We then have a command line argument to control which language is being tested, of the form:

          -- language python/matlab

(the default for now being `matlab` as controlled by this line: https://github.com/Cambridge-ICCS/ONEFlux/blob/unifiedEngine/tests/conftest.py#L38)

e.g.,

```
python3.11 -m pytest tests/unit_tests/test_ustar_cp/test_fcEqnAnnualSine.py --language python
```
This will run the python test for `test_fcEqnAnnualSine.py`.

### CODE REVIEW

**Comment by j-emberton** [comment]
This looks great Dom.

I can see you use a linter/formatter that removes trailing whitespaces from Python code. Shall I make a quick git action to add a hook/action to make this consistent for all of us?

**Comment by dorchard** [comment]
Sorry that is a bit of a quirk of my IDE setup; but setting a hook to do it might be an idea if people don't mind (or I turn it off in my IDE).

**Comment by j-emberton** [comment]
I think setting up a ruff pre-commit hook would be useful given we're about to do a lot more python work imminently :) I'll do it now, but will set it up to only work on our stuff not the full repo.

Added as issue #65

File: tests/conftest.py [hunk 1]
```diff
@@ -28,9 +28,47 @@
...
+
+    def equal(self, x, y) -> bool:
+        return x == y
+
+    # Overload calling of methods and 'rethrow' to global context
+    def __getattribute__(self,name):
+        def newfunc(*args, **kwargs):
+          result = eval(name + "(*" + str(args) + ", *" + str(kwargs) + ")")
+          return result
+        return newfunc
```
Comment (isaacaka) [inline tests/conftest.py]:
```suggestion
  def __getattribute__(self, name):
      def newfunc(*args, **kwargs):
          func = globals().get(name)  # Access global dictionary of defined functions
          if callable(func):
              return func(*args, **kwargs)
          raise AttributeError(f"'{name}' is not callable")
      return newfunc
```
This way can avoid the use of eval.

**Review by isaacaka — COMMENTED** [review]
Nice, I made a code suggestion to avoid the use of eval.

**Comment by dorchard** [comment]
> Nice, I made a code suggestion to avoid the use of eval.

Ah brilliant - thanks!

**Comment by j-emberton** [comment]
Hi Dom, what's the status on this? Is it ready to start deploying on our already existing tests?

**Comment by dorchard** [comment]
Yes I think so - but we're going to need to rewrite our existing tests to use its `convert` and `compare` so this is currently marked as draft as not all the tests have been converted over. I think we should do that here then merge with `ustar_cp_refactor_main`.

**Comment by j-emberton** [comment]
OK, I'd like to help out if you want. What could I look at first?

**Comment by dorchard** [comment]
That would be great. I think it's a matter of going through the existing tests and replacing occurences of `matlab.double` and ` to_matlab_type` calls with `test_engine.compare` and `compare_matlab_arrays` with `test_engine.equal` (a function which takes two arguments)

**Comment by j-emberton** [comment]
This PR is now fully updated to be in sync with `ustar_cp_refactor_main` as of 16:30, 12 Dec.

All tests use new runner and pass.

**Comment by dorchard** [comment]
I have done some work on this to allow us to switch languages more easily now via the command line. This seems to be working okay for `test_fcEqnAnnualSine.py` which currently has just a stub in the python (which will fail all the tests, but they do get rune).

**Comment by j-emberton** [comment]
I've moved over some tests from `matlab_engine` to `test_engine`, so now all 'matlab' tests pass.

I think there's still an unresolved problem in that the `PythonEngine` 'convert' method doesn't have any way of understanding what the required data type for the python function is.

To date we have used lists as a proxy for matlab arrays, however in python we will need to convert these lists to numpy arrays on some occasions and keep them as lists on others (The `stats` data structure is a list of dicts in python for instance).

**Comment by dorchard** [comment]
Great thanks. 
That makes sense. Perhaps we should do that on demand then as we move to translation phase and can fix up `convert` on demand? Or I could have an initial stab first that numpy array converts- but it might be easier to do with concrete tests being run to program against?

**Comment by j-emberton** [comment]
Yeah ok. I propose we merge this then.

**Comment by isaacaka** [comment]
I had a look at the merge, I agree with the on demand merges. In the case of running the python engine. 

Results also would need to be converted to matlab data type:
 `assert compare_matlab_arrays(to_matlab_type(result), to_matlab_type(expected))`
 Thus running `to_matlab_type` on matlab data types should be idempotent

**Pull request merged**